client/beefy/src/communication/request_response/outgoing_requests_engine.rs

```diff
@@ -0,0 +1,246 @@
+	service::NetworkRequest,
+};
+use sp_api::ProvideRuntimeApi;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block, NumberFor},
+};
+use std::{collections::VecDeque, result::Result, sync::Arc};
+
+use crate::{
+	communication::request_response::{Error, JustificationRequest},
+	justification::{decode_and_verify_finality_proof, BeefyVersionedFinalityProof},
+	KnownPeers,
+};
+
+/// Response type received from network.
+type Response = Result<Vec<u8>, RequestFailure>;
+/// Used to receive a response from the network.
+type ResponseReceiver = oneshot::Receiver<Response>;
+
+enum State<B: Block> {
+	Idle(stream::Pending<Result<Response, Canceled>>),
+	AwaitingResponse(PeerId, NumberFor<B>, stream::Once<ResponseReceiver>),
+}
+
+pub struct OnDemandJustificationsEngine<B: Block, N, R> {
+	network: N,
+	runtime: Arc<R>,
+	protocol_name: ProtocolName,
+
+	live_peers: Arc<Mutex<KnownPeers<B>>>,
+	peers_cache: VecDeque<PeerId>,
+
+	state: State<B>,
+}
+
+impl<B, N, R> OnDemandJustificationsEngine<B, N, R>
+where
+	B: Block,
+	N: NetworkRequest,
+	R: ProvideRuntimeApi<B>,
+	R::Api: BeefyApi<B>,
+{
+	pub fn new(
+		network: N,
+		runtime: Arc<R>,
+		protocol_name: ProtocolName,
+		live_peers: Arc<Mutex<KnownPeers<B>>>,
+	) -> Self {
+		Self {
+			network,
+			runtime,
+			protocol_name,
+			live_peers,
+			peers_cache: VecDeque::new(),
+			state: State::Idle(stream::pending()),
+		}
+	}
+
+	fn reset_peers_cache_for_block(&mut self, block: NumberFor<B>) {
+		// TODO (issue #12296): replace peer selection with generic one that involves all protocols.
+		self.peers_cache = self.live_peers.lock().at_least_at_block(block);
+	}
+
+	fn try_next_peer(&mut self) -> Option<PeerId> {
+		// TODO (issue #12296): replace peer selection with generic one that involves all protocols.
+		let live = self.live_peers.lock();
+		while let Some(peer) = self.peers_cache.pop_front() {
+			if live.contains(&peer) {
+				return Some(peer)
+			}
+		}
+		None
+	}
+
+	fn request_from_peer(&mut self, peer: PeerId, block: NumberFor<B>) {
+		debug!(target: "beefy::sync", "🥩 requesting justif #{:?} from peer {:?}", block, peer);
+
+		let payload = JustificationRequest::<B> { begin: block }.encode();
+
+		let (tx, rx) = oneshot::channel();
+
+		self.network.start_request(
+			peer,
+			self.protocol_name.clone(),
+			payload,
+			tx,
+			IfDisconnected::ImmediateError,
+		);
+
+		self.state = State::AwaitingResponse(peer, block, stream::once(rx));
+	}
+
+	/// If no other request is in progress, start new justification request for `block`.
+	pub fn request(&mut self, block: NumberFor<B>) {
+		// ignore new requests while there's already one pending
+		match &self.state {
+			State::AwaitingResponse(_, _, _) => return,
+			State::Idle(_) => (),
+		}
+		self.reset_peers_cache_for_block(block);
+
+		// Start the requests engine - each unsuccessful received response will automatically
+		// trigger a new request to the next peer in the `peers_cache` until there are none left.
```

Comment:
One thing I was wondering is that since we only send one request at a time, can the remote node (either maliciously or not) cause denial of service in our end by not responding? I didn't see any code explicitly handling the case where the request timeouts so I'm wondering is that error merged with `RequestChannelExhausted`/`oneshot::Canceled` or is there some other mechanism that informs the outgoing BEEFY request engine that a request timed out.

Reply:
Yeah, the underlying `Network` object takes care of this for us. When configuring the protocol we also configure a `request_timeout` [here](https://github.com/acatangiu/substrate/blob/68854a866da83e88caa8c939a3b69cc5ec2c470e/client/beefy/src/communication/request_response/mod.rs#L63).

Reply:
Yeah I saw the constant but was confused because I didn't see the code that explicitly handles request timeouts. But another look in `sc-network`'s request-response code cleared up the the issue.